_posts/2024-06-24-solarkraft-part4.md

```diff
@@ -0,0 +1,200 @@
+
+
+Can the described approach of direct monitors be considered satisfactory? Please stop to think about it for a sec, before opening our answer below.
+
+<details>
+<summary> <b>Are direct monitors enough?</b></summary>
+
+<p>You may have guessed the answer: we believe NO! And here is an (incomplete) list of whys:</p>
+
+<ol>
+<li> A method may have a side effect, which was overlooked by the spec author. E.g. a boolean flag is set, which in another method allows to direct funds to another account.</li>
+<li> Code evolves, but the spec stays as is; as a result a side effect above is introduced unintentionally, with the stale spec not accounting for it.</li>
+<li> Internal state component is modified in multiple methods, in different ways. The specification about how the component should be updated is scattered in multiple places, and loopholes may easily creep in.</li>
+<li> An invariant which is preserved by the method of this contract, is violated by a method from another contract. As no specs are written or monitored for this other contract, no violation is detected.</li>
+</ol>
+
+What is fundamentally missing from direct monitors, is the guarantee of <i>completeness</i>. Thus, we proceed to explore <i>reverse monitors</i>, which are complementary to direct monitors, specifically for ensuring specification completeness.
+</details>
+
+
+## Reverse Monitors
+
+With reverse reasoning we will try to patch the loopholes that were left by direct monitor specs above. To do so, we start with an _effect_ (state was modified), and go back to its _cause_ (what should have happened taking the effect into account). Here is the corresponding picture which puts a bit of structure into the reverse reasoning.
+
+![Reverse monitor specs](/img/ReverseMonitors.png)
+
+
+- `MonitorTrigger_i` is a condition which triggers (activates) the monitor. If _any_ of those conditions hold, the monitor is activated;
+- `MonitorEffect_i` is a condition over past and next state variables, which specifies the effect that the the monitor, if activated, should check. _All_ of `MonitorEffect_i` should hold if the transaction is successful.
+
+Similar to direct reasoning, in the above, `Monitor<Trigger|Effect>` is a prefix, which tells the monitor system how to interpret this predicate. The complete pattern for predicate names with these prefixes is as follows:
+
+```
+Monitor<Trigger|Effect>_<Monitor>_<ConditionDescription>
+```
+
+All predicates which refer to the same `<Monitor>` will be grouped, to create together one _effect monitor_. Interpreted formally, the monitor should do the following when activated:
+
+- If _any_ of `MonitorTrigger_i` conditions fire, check that _all_ `MonitorEffect_i` hold over the past and next states (otherwise, issue a warning / revert if configured to do so)
```

Comment:
I don't remember if I made this comment on your original writeup, but how exactly is this different from `MustPass_ & MustHold_`? Even if it is, you really should describe the difference explicitly.

Reply:
I have added more explanations; thanks for pointing this out!